teps/0137-cloudevents-controller.md

```diff
@@ -0,0 +1,527 @@
+  - [Test Plan](#test-plan)
+  - [Infrastructure Needed](#infrastructure-needed)
+  - [Upgrade and Migration Strategy](#upgrade-and-migration-strategy)
+  - [Implementation Pull Requests](#implementation-pull-requests)
+- [References](#references)
+<!-- /toc -->
+
+## Summary
+
+Tekton supports [sending CloudEvents][tekton-ce-docs] for events associated
+with `TaskRuns`, `PipelineRuns` and `CustomRuns`. Events associated with the
+latter are generated by a dedicated controller binary, while the former two
+are generated by the `TaskRun` and `PipelineRun` controller respectively.
+
+In this TEP we propose to move all the CloudEvents generation logic to the
+separate controller. The new controller will support new versions and types
+of events:
+
+- Tekton v2 events, which embed `v1.TaskRun` and `v1.PipelineRun` as opposed
+  to v1 events which embed the corresponding `v1beta1` resources
+- [CDEvents core events][cdevents-core]
+- User driven [CDEvents][cdevents] of any kind
+
+## Motivation
+
+When Tekton introduced support for events, the [CloudEvents][cloudevents]
+format was chosen as it provides an interoperable standard for the format of
+the enveloper of events. At the time, no such standard existed for the
+payload. Because of that I created, together with other people on the
+[CDF Interoperability Special Interest Group][cdf-sig-interop], the
+[CDEvents][cdevents] project. The TEP proposes to adopt that standard, now that
+it is available, to enhance Tekton interoperability with other tools in the
+CI/CD space.
+
+Rather then introducing CDEvents support in the core Tekton controllers, this
+TEP additionally proposes to split the event feature into a dedicated controller,
+work which has already started for the `CustomRun` events, and which provides a
+better separation of concerns and security benefits, as it separates the secrets
+which are needed by `TaskRun` and `PipelineRun` controllers from those required
+to successfully deliver events to sinks.
+
+Finally, the new controller enables the development of new features in the area
+of events and notification with no or reduced impact on the `TaskRun` and `PipelineRun`
+controller, which means the code in those controller remains leaner, and their
+scalability is not compromised.
+
+### Goals
+
+- Produce Tekton v2 events through the dedicate event controller
+- Produce core CDEvents through the dedicate event controller
+- Give task authors a way to trigger additional events from their tasks. Tekton
+  has no knowledge about abstractions like builds, artifacts, tests and more,
+  that are modelled by CDEvents. Tekton Task authors write task that implement
+  such abstraction. Goal of this TEP is to let such authors notify about them
+  through CDEvents.
+- Give users the ability to select through configuration which event type they
+  would like to send, if any
+- Give Tekton operators the ability to remove event support from their
+  Tekton deployment
```

Comment:
Does this mean the event support is enabled by default?

Reply:
No, events are only sent if a sink is configured, exactly like today.
Today every reconcile cycle checks the configuration to see if events need to be sent. 
That logic will be moved to a dedicated controller and will grow feature-wise over time. Operators may decide to not install the event controller at all if they do not plan to use events.